onnx/defs/traditionalml/defs.cc

```diff
@@ -19,6 +19,22 @@ ONNX_ML_OPERATOR_SET_SCHEMA(
         .Input(0, "X", "Data to be selected", "T")
         .Input(1, "Y", "The indices, based on 0 as the first index of any dimension.", "tensor(int64)")
         .Output(0, "Z", "Selected output data as an array", "T")
+        .TypeAndShapeInferenceFunction([](InferenceContext& ctx) {
+          propagateElemTypeFromInputToOutput(ctx, 0, 0);
+          if (!hasNInputShapes(ctx, 1)) {
+            return;
+          }
+          const auto& input_shape = ctx.getInputType(0)->tensor_type().shape();
+          const auto input_ndim = input_shape.dim_size();
+
+          auto output_shape = ctx.getOutputType(0)->mutable_tensor_type()->mutable_shape();
+          // This operator only applies to the last dimension; thus -1
+          for (int i = 0; i < input_ndim - 1; ++i) {
+            *output_shape->add_dim() = input_shape.dim(i);
+          }
+          // The length of the last dimension is not statically known
+          output_shape->add_dim();
```

Comment:
This looks correct. I think we could make it more precise if the second tensor's dimension-size is statically known, I suppose.

Reply:
Specifically, I think we could do something like
```cpp
   if (hasInputShape(ctx, 1)) {
      const auto& indices_shape = getInputShape(ctx, 1);
      if (indices_shape.dim_size() > 0) {
         *output_shape->add_dim() = indices_shape.dim(0)
      }
   }
```

Reply:
with the existing code for the else-cases

Reply:
However, I notice that the op's description doesn't mention the order of axes for the output. This looks like a missing piece of information.

Reply:
onnxruntime, however, seems to be consistent with the above ordering.

Reply:
While missing in the specification, this appears to me to be the behavior people would expect. I added your suggestion to the shape inference. Thanks!